src/overlays/actors/ovl_Bg_Jya_Megami/z_bg_jya_megami.c

```diff
@@ -21,31 +29,325 @@ const ActorInit Bg_Jya_Megami_InitVars = {
+                sp8C.z += 10.0f;
+                func_8089A1DC(globalContext, &sp8C, &temp->vel, 3, 2);
+            }
+        } else if (this->explosionTimer == temp2->delay) {
+            Math_Vec3f_Sum(&temp->pos, &temp2->unk_00, &sp8C);
+            sp8C.z += 10.0f;
+            func_8089A1DC(globalContext, &sp8C, &temp->vel, 4, 2);
+        }
+    }
+
+    if ((this->explosionTimer % 4 == 0) && (this->explosionTimer > 30) && (this->explosionTimer < 80) &&
+        (this->explosionTimer > 40)) {
+        sp8C.x = ((Math_Rand_ZeroOne() - 0.5f) * 90.0f) + this->dyna.actor.posRot.pos.x;
+        sp8C.y = (this->dyna.actor.posRot.pos.y - (Math_Rand_ZeroOne() * 80.0f)) - 20.0f;
+        sp8C.z = this->dyna.actor.posRot.pos.z - ((Math_Rand_ZeroOne() - 0.5f) * 50.0f);
+        func_8089A1DC(globalContext, &sp8C, &sVec, 1, 0);
+    }
+    if (this->explosionTimer < ARRAY_COUNT(sPiecesInit)) {
+        sp8C.x = this->dyna.actor.posRot.pos.x;
+        sp8C.y = this->dyna.actor.posRot.pos.y - 60.0f;
+        sp8C.z = this->dyna.actor.posRot.pos.z;
+        func_80033480(globalContext, &sp8C, 100.0f, 1, 150, 100, 1);
+    }
+    if (this->explosionTimer == 60) {
+        func_80078884(NA_SE_SY_CORRECT_CHIME);
+    }
+    if (this->explosionTimer >= 100) {
+        Actor_Kill(&this->dyna.actor);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Jya_Megami/func_8089A41C.s")
+void BgJyaMegami_Update(Actor* thisx, GlobalContext* globalContext) {
+    BgJyaMegami* this = THIS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Jya_Megami/BgJyaMegami_Init.s")
+    this->actionFunc(this, globalContext);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Jya_Megami/BgJyaMegami_Destroy.s")
+void BgJyaMegami_DrawDetectLight(BgJyaMegami* this, GlobalContext* globalContext) {
+    GraphicsContext* gfxCtx;
+    Gfx* dispRefs[4];
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Jya_Megami/func_8089A5EC.s")
+    gfxCtx = globalContext->state.gfxCtx;
+    Graph_OpenDisps(dispRefs, globalContext->state.gfxCtx, "../z_bg_jya_megami.c", 706);
+    func_80093D18(globalContext->state.gfxCtx);
+    gSPSegment(gfxCtx->polyOpa.p++, 0x08, SEGMENTED_TO_VIRTUAL(sRightSideCrumbles[this->crumbleIndex]));
+    gSPSegment(gfxCtx->polyOpa.p++, 0x09, SEGMENTED_TO_VIRTUAL(sLeftSideCrumbles[this->crumbleIndex]));
+    gSPMatrix(gfxCtx->polyOpa.p++, Matrix_NewMtx(globalContext->state.gfxCtx, "../z_bg_jya_megami.c", 716),
+              G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+    gSPDisplayList(gfxCtx->polyOpa.p++, &D_06005780);
+    Graph_CloseDisps(dispRefs, globalContext->state.gfxCtx, "../z_bg_jya_megami.c", 720);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Jya_Megami/func_8089A604.s")
+void BgJyaMegami_DrawExplosion(BgJyaMegami* this, GlobalContext* globalContext) {
+    GraphicsContext* gfxCtx;
+    BgJyaMegami_Piece* temp;
+    u32 i;
+    Gfx* dispRefs[5];
```

Comment:
since we know open disps is part of a macro, it would be helpful if you make it work like the macro does while we wait for the macro to be implemented.
The pattern is that dispRefs should always be size 4. You may need to add a pad. And since it gets declared below the normal block at the top of the function, it should be last in your declaration order. With gfxCtx being on top.
My guess would be something like:

    s32 pad;
    BgJyaMegami_Piece* temp;
    u32 i;
    GraphicsContext* gfxCtx;
    Gfx* dispRefs[4];

Reply:
and maybe just name the temp piece

Reply:
didnt know about this macro part, but makes sense